File: src/backend/ISeeSharp.Domain/Entities/Session.cs
namespace ISeeSharp.Domain.Entities;

public class Session : Entity
{
    public string Title { get; private set; } = string.Empty;
    public string Slug { get; private set; } = string.Empty;
    public string Description { get; private set; } = string.Empty;
    public string Instructions { get; private set; } = string.Empty;
    public string? ObjectivesJson { get; private set; }
    public string? HintsJson { get; private set; }

    public SessionDifficulty Difficulty { get; private set; }
    public SessionCategory Category { get; private set; }

    public int Points { get; private set; }
    public int EstimatedMinutes { get; private set; }
    public int TargetWpm { get; private set; }

    public bool IsActive { get; private set; } = true;
    public bool IsPremium { get; private set; }

    public int TotalAttempts { get; private set; }
    public int TotalCompletions { get; private set; }
    public double AverageAccuracy { get; private set; }
    public double AverageWpm { get; private set; }

    private readonly List<SessionFile> _files = [];
    public IReadOnlyCollection<SessionFile> Files => _files.AsReadOnly();

    private readonly List<UserSessionResult> _userResults = [];
    public IReadOnlyCollection<UserSessionResult> UserResults => _userResults.AsReadOnly();

    public int TotalLines => _files.Sum(f => f.TotalLines);
    public int TotalCharacters => _files.Sum(f => f.TotalCharacters);

    private Session() { }

    public static Session Create(
        string title,
        string slug,
        string description,
        string instructions,
        SessionDifficulty difficulty,
        SessionCategory category,
        int points,
        int estimatedMinutes = 10,
        int targetWpm = 40,
        bool isPremium = false,
        string? objectivesJson = null,
        string? hintsJson = null)
    {
        return new Session
        {
            Title = title,
            Slug = slug,
            Description = description,
            Instructions = instructions,
            Difficulty = difficulty,
            Category = category,
            Points = points,
            EstimatedMinutes = estimatedMinutes,
            TargetWpm = targetWpm,
            IsPremium = isPremium,
            ObjectivesJson = objectivesJson,
            HintsJson = hintsJson
        };
    }

    public void AddFile(SessionFile file)
    {
        _files.Add(file);
        UpdatedAt = DateTime.UtcNow;
    }

    public void IncrementAttempt()
    {
        TotalAttempts++;
        UpdatedAt = DateTime.UtcNow;
    }

    public void RecordCompletion(double accuracy, int wpm)
    {
        var totalCompleted = TotalCompletions + 1;
        AverageAccuracy = ((AverageAccuracy * TotalCompletions) + accuracy) / totalCompleted;
        AverageWpm = ((AverageWpm * TotalCompletions) + wpm) / totalCompleted;
        TotalCompletions = totalCompleted;
        UpdatedAt = DateTime.UtcNow;
    }

    public void Deactivate()
    {
        IsActive = false;
        UpdatedAt = DateTime.UtcNow;
    }

    public void Activate()
    {
        IsActive = true;
        UpdatedAt = DateTime.UtcNow;
    }
}

public enum SessionDifficulty
{
    Easy = 0,
    Medium = 1,
    Hard = 2,
    Expert = 3
}

public enum SessionCategory
{
    Syntax = 0,
    Variables = 1,
    ControlFlow = 2,
    Methods = 3,
    Classes = 4,
    Interfaces = 5,
    Inheritance = 6,
    Generics = 7,
    Collections = 8,
    Linq = 9,
    AsyncAwait = 10,
    ErrorHandling = 11,
    Patterns = 12,
    ModernCSharp = 13,
    FullProject = 14
}

File: src/frontend/types/session.ts
export type SessionDifficulty = "Easy" | "Medium" | "Hard" | "Expert";

export type SessionCategory =
  | "Syntax"
  | "Variables"
  | "ControlFlow"
  | "Methods"
  | "Classes"
  | "Interfaces"
  | "Inheritance"
  | "Generics"
  | "Collections"
  | "Linq"
  | "AsyncAwait"
  | "ErrorHandling"
  | "Patterns"
  | "ModernCSharp"
  | "FullProject";

export interface SessionFile {
  id: string;
  sessionId: string;
  path: string;
  displayName: string;
  language: string;
  targetContent: string;
  editableRegionsJson: string | null;
  sortOrder: number;
  totalLines: number;
  totalCharacters: number;
}

export interface Session {
  id: string;
  title: string;
  slug: string;
  description: string;
  instructions: string;
  objectivesJson: string | null;
  hintsJson: string | null;
  difficulty: SessionDifficulty;
  category: SessionCategory;
  points: number;
  estimatedMinutes: number;
  targetWpm: number;
  isActive: boolean;
  isPremium: boolean;
  totalAttempts: number;
  totalCompletions: number;
  averageAccuracy: number;
  averageWpm: number;
  totalLines: number;
  totalCharacters: number;
  files: SessionFile[];
}

export interface SessionListItem {
  id: string;
  title: string;
  slug: string;
  description: string;
  difficulty: SessionDifficulty;
  category: SessionCategory;
  points: number;
  estimatedMinutes: number;
  targetWpm: number;
  totalLines: number;
  totalFiles: number;
  totalAttempts: number;
  totalCompletions: number;
  averageAccuracy: number;
}

export interface EditableRegion {
  startLine: number;
  startColumn: number;
  endLine: number;
  endColumn: number;
  hint?: string;
}

export function parseEditableRegions(json: string | null): EditableRegion[] {
  if (!json) return [];
  try {
    return JSON.parse(json);
  } catch {
    return [];
  }
}

export function parseObjectives(json: string | null): string[] {
  if (!json) return [];
  try {
    return JSON.parse(json);
  } catch {
    return [];
  }
}

export function parseHints(json: string | null): string[] {
  if (!json) return [];
  try {
    return JSON.parse(json);
  } catch {
    return [];
  }
}
